packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -2,199 +2,109 @@ import * as fs from 'fs-extra'
-          )
-        page.staticData[key] = staticData
-      }
-
-      if (currentFile) {
-        let dataKeys = currentFile.dataKeys.get(pageId)
-        if (!dataKeys) currentFile.dataKeys.set(pageId, (dataKeys = []))
-        dataKeys.push(key)
-      }
-
-      if (existingPage) {
-        changedPages.add(pageId)
-        emitChange()
-      }
-    }
-
-    function removePageData(dataKeys: string[], pageId: string) {
-      const page = pageCache[pageId]
-
-      dataKeys.forEach((key) => {
-        delete page.data[key]
-        delete page.staticData[key]
-      })
-
-      if (hasKeys(page.data) || hasKeys(page.staticData)) {
-        changedPages.add(pageId)
-        emitChange()
-      } else {
-        delete pageCache[pageId]
-      }
-    }
-
-    function hasKeys(obj: object) {
-      for (const _ in obj) return true
-      return false
-    }
-
-    function watchFiles(
-      baseDir: string,
-      arg2?: string | string[] | FileHandler,
-      arg3?: FileHandler
-    ) {
-      let globs: string[]
-      let handler: FileHandler
-
-      if (typeof arg2 === 'function') {
-        globs = ['**/*']
-        handler = arg2
-      } else {
-        globs = Array.isArray(arg2) ? arg2 : [arg2 || '**/*']
-        handler =
-          arg3 || ((file) => Promise.resolve(loadPageData(file, helpers)))
-      }
-
+    function watchFiles({
+      baseDir,
+      globs = ['**/*'],
+      fileHandler = defaultFileHandler,
+      ignored = [],
+    }: {
+      baseDir: string
+      globs?: string | string[]
+      fileHandler?: FileHandler
+      ignored?: string[]
+    }) {
       // Strip trailing slash and make absolute
       baseDir = path.resolve(pagesDir, baseDir)
 
-      const processFile = async (
-        type: 'add' | 'change' | 'unlink',
-        file: File
-      ) => {
-        file.dataKeys.forEach(removePageData)
-        if (type !== 'unlink') {
-          if (type === 'change') {
-            file.content = null
-            file.dataKeys.clear()
-          }
-          currentFile = file
-          try {
-            const pageData = await handler(file)
-            if (pageData) {
-              addPageData(pageData)
-            }
-          } finally {
-            currentFile = null
-          }
-        }
-      }
-
       watchers.add(
         chokidar
           .watch(globs, {
             cwd: baseDir,
-            ignored: ['**/node_modules/**/*'],
+            ignored: ['**/node_modules/**/*', '**/.git/**', ...ignored],
           })
-          .on('all', (type, filePath) => {
-            if (type !== 'addDir' && type !== 'unlinkDir') {
-              filePath = path.join(baseDir, filePath)
-
-              const file =
-                fileCache[filePath] ||
-                (fileCache[filePath] = new File(filePath, baseDir))
+          .on('add', handleFileChange)
+          .on('change', handleFileChange)
+          .on('unlink', handleFileUnLink)
```

Comment:
Can you explain the need to change the file watching logic? I thought very carefully when implementing it, and I'm worried you may be introducing bugs.

Reply:
To enable the proxy api, I have to track the associations between file and page carefully. So I move this logic into a dedicated class `PagesDataKeeper` for maintainability.
I have tried my best to transplant all your logic in to the new implementation. Please let me know if I missed somthing!

Reply:
This fileSystem<->pageData mapping thing is indeed very tricky. That also why I use more classes and data structures to represent the model.